### PR TITLE
K8s: fixed issues related to links / ownership links

### PR DESCRIPTION
**Pull request by hunchback**
Make improvements and fixes to the following links:
- pod --ownership--> node
- container --ownership--> pod
- container --association--> docker
- networkPolicy --association--> pod

### CODE REVIEW

**Comment by hunchback** [comment]
@lebauce / @safchain - kindly review ...

File: topology/probes/k8s/client.go [hunk 1]
```diff
@@ -80,6 +81,41 @@ type kubeCache struct {
...
+	return c.cache.List()
+}
+
+func getStringField(obj interface{}, field string) string {
+	r := reflect.ValueOf(obj)
+	f := reflect.Indirect(r).FieldByName(field)
+	return f.String()
+}
+
+func (c *kubeCache) listByNamespace(namespace string) (objList []interface{}) {
```
Comment (lebauce) [inline topology/probes/k8s/client.go]: It seems this function is used only once, and the type is perfectly known so there is no reason to use a slow reflection to get a simple field.

File: topology/probes/k8s/graph.go [hunk 2]
```diff
@@ -24,22 +24,58 @@ package k8s
 
 import (
 	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/topology"
 	"github.com/skydive-project/skydive/topology/graph"
 )
 
+func setManager(m *graph.Metadata) {
```
Comment (lebauce) [inline topology/probes/k8s/graph.go]: Maps are passed as reference so there's not need to pass it as a pointer.

Besides, it's a bit overkill to create a function just to set an attribute. Maybe defining a constant for the manager is enough (KubernetesManager ?) and set it inline where setManager is called.

File: topology/probes/k8s/container.go [hunk 3]
```diff
@@ -86,16 +86,35 @@ func containerUID(pod *v1.Pod, containerName string) graph.Identifier {
 	return graph.GenIDNameBased(string(pod.GetUID()), containerName)
 }
 
+func dumpContainer2(podNamespace, podName, containerName string) string {
```
Comment (lebauce) [inline topology/probes/k8s/container.go]: `func dumpContainer(pod *v1.Pod, containerName string) string` may be enough

File: topology/probes/k8s/client.go (hunk 1 above)
Comment (hunchback) [reply topology/probes/k8s/client.go]: Is is going to be used by other throughout the code base once we create graph.Node to represent namespaces as in:

```
$ kubectl get namespaces
NAME          STATUS    AGE
default       Active    3d
kube-public   Active    3d
kube-system   Active    3d
```

File: topology/probes/k8s/container.go (hunk 3 above)
Comment (hunchback) [reply topology/probes/k8s/container.go]: I wanted to have a dump for containers when we don't have readily available kubernetes objects (such as v1.Pod, etc.)

File: topology/probes/k8s/graph.go (hunk 2 above)
Comment (hunchback) [reply topology/probes/k8s/graph.go]: agreed x 2 ... will fix accordingly

File: topology/probes/k8s/container.go (hunk 3 above)
Comment (lebauce) [reply topology/probes/k8s/container.go]: For the only call to dumpContainer2 (line 175 of container.go), we have v1.Pod object

File: topology/probes/k8s/client.go (hunk 1 above)
Comment (lebauce) [reply topology/probes/k8s/client.go]: I suggest we do this modification when we'll really need it.

File: topology/probes/k8s/networkpolicy.go [hunk 4]
```diff
@@ -46,134 +48,111 @@ func (n *networkPolicyProbe) newMetadata(np *networking_v1.NetworkPolicy) graph.
...
-		} else if policy.Namespace == api.NamespaceAll || pod.Namespace == policy.Namespace {
-			toLink = true
-		}
-
-		if toLink {
-			n.graph.Link(policyNode, podNode, policyToPodMetadata)
-		} else {
-			n.graph.Unlink(policyNode, podNode)
-		}
+		syncLink(n.graph, policyNode, podNode, n.isPodSelected(policy, pod))
```
[the remark this answers is not in the document]
Comment (hunchback) [reply topology/probes/k8s/networkpolicy.go]: Have fixed syncLink so now it either added a link or deletes a link (thus reverting to the original behavior of the code).

Please see commit named: `k8s: add utility functions for handling graph.Edge`

File: topology/probes/k8s/container.go (hunk 3 above)
Comment (hunchback) [reply topology/probes/k8s/container.go]: ok, will be addressed in: `k8s: improve logging of CRUD operations`

File: topology/probes/k8s/client.go (hunk 1 above)
Comment (hunchback) [reply topology/probes/k8s/client.go]: ok, will be addressed in update to: `k8s: networkPolicy chain namespace-filter then pod-selection-filter`

File: topology/topology.go [hunk 5]
```diff
@@ -119,21 +119,36 @@ func IsOwnershipLinked(g *graph.Graph, node *graph.Node) bool {
 	return len(edges) != 0
 }
 
-// AddOwnershipLink Link between the parent and the child node, the child can have only one parent, previous will be overwritten
-func AddOwnershipLink(g *graph.Graph, parent *graph.Node, child *graph.Node, metadata graph.Metadata) *graph.Edge {
-	// a child node can only have one parent of type ownership, so delete the previous link
-	for _, e := range g.GetNodeEdges(child, graph.Metadata{"RelationType": OwnershipLink}) {
+// GetLink get Link between the parent and the child node or nil
+func GetLink(g *graph.Graph, parent *graph.Node, child *graph.Node, metadata graph.Metadata) *graph.Edge {
```
[the remark this answers is not in the document]
Comment (hunchback) [reply topology/topology.go]: ok, will address in update to: `k8s: add utility functions for handling graph.Edge`

File: topology/probes/k8s/networkpolicy.go [hunk 6]
```diff
@@ -46,134 +48,111 @@ func (n *networkPolicyProbe) newMetadata(np *networking_v1.NetworkPolicy) graph.
...
-	if policy, ok := obj.(*networking_v1.NetworkPolicy); ok {
-		n.graph.Lock()
-		policyNode := n.graph.NewNode(netpolUID(policy), n.newMetadata(policy))
-		n.handleNetworkPolicy(policyNode, policy)
-		n.graph.Unlock()
-	}
+	n.handleNetworkPolicyAdd(obj)
 }
 
 func (n *networkPolicyProbe) OnUpdate(oldObj, newObj interface{}) {
```
[the remark this answers is not in the document]
Comment (hunchback) [reply topology/probes/k8s/networkpolicy.go]: ok have returned the prior behavior, it was replaced not sue to a specific but but so as to simplify the logic, but as said now reverted

**Comment by hunchback** [comment]
@lebauce - I have uploaded a new version fixing all identified issues - please review

**Comment by hunchback** [comment]
@lebauce, @safchain - I have addressed setting hostID to "" (for Node/Edge elements):

```
topology: add ability to set host in g.Link() AddOwnershipLink()
k8s: add/fix newNode(), link(), addOwnershipLink() with empty hostID
k8s: call new newNode() which sets emtpry hostID
flow: fixed bad gofmt
```

**Comment by safchain** [comment]
ok to test

**Comment by lebauce** [comment]
add to whitelist

**Comment by safchain** [comment]
rerun functional-tests-backend-elasticsearch

**Pull request merged**